cli/azd/pkg/pipeline/pipeline_manager.go

```diff
@@ -126,6 +128,87 @@ func (pm *PipelineManager) ScmProviderName() string {
 	return pm.scmProvider.Name()
 }
 
+type servicePrincipalResult struct {
+	appIdOrName      string
+	applicationName  string
+	lookupKind       servicePrincipalLookupKind
+	servicePrincipal *graphsdk.ServicePrincipal
+}
+
+func servicePrincipal(
+	ctx context.Context,
+	envClientId,
+	subscriptionId string,
+	args *PipelineManagerArgs, adService azcli.AdService) (*servicePrincipalResult, error) {
+	// Existing Service Principal Lookup strategy
+	// 1. --principal-id
+	// 2. --principal-name
+	// 3. AZURE_PIPELINE_CLIENT_ID environment variable
+	// 4. Create new service principal with default naming convention
+	var appIdOrName, applicationName string
+	var lookupKind servicePrincipalLookupKind
+
+	if args.PipelineServicePrincipalId != "" {
+		appIdOrName = args.PipelineServicePrincipalId
+		lookupKind = lookupKindPrincipalId
+	} else if args.PipelineServicePrincipalName != "" {
+		appIdOrName = args.PipelineServicePrincipalName
+		lookupKind = lookupKindPrincipleName
+	} else if envClientId != "" {
+		appIdOrName = envClientId
+		lookupKind = lookupKindEnvironmentVariable
+	}
+
+	if appIdOrName != "" {
```

Comment:
Consider inverting this if block to reduce nesting in the code.

Reply:
love it!.  Done